node/pkg/watchers/hedera/watcher.go

```diff
@@ -0,0 +1,612 @@
+//       "internalType": "bytes",
+//       "name": "payload",
+//       "type": "bytes"
+//     },
+//     {
+//       "indexed": false,
+//       "internalType": "uint8",
+//       "name": "consistencyLevel",
+//       "type": "uint8"
+//     }
+//   ],
+//   "name": "LogMessagePublished",
+//   "type": "event"
+// },
+
+type (
+	// Watcher is responsible for looking over a cosmwasm blockchain and reporting new transactions to the contract
+	Watcher struct {
+		// urlWS    string // gRPC websocket URL
+		urlLCD   string // REST URL
+		contract string // topic
+
+		msgChan chan *common.MessagePublication
+
+		// Incoming re-observation requests from the network. Pre-filtered to only
+		// include requests for our chainID.
+		obsvReqC chan *gossipv1.ObservationRequest
+
+		// Readiness component
+		readiness readiness.Component
+		// VAA ChainID of the network we're connecting to.
+		chainID vaa.ChainID
+		// // Key for contract address in the wasm logs
+		// contractAddressFilterKey string
+		// // Key for contract address in the wasm logs
+		// contractAddressLogKey string
+
+		// URL to get the latest block info from
+		// latestBlockURL string
+	}
+)
+
+// NewWatcher creates a new hedera watcher
+func NewWatcher(
+	// urlWS string,
+	urlLCD string,
+	contract string,
+	lockEvents chan *common.MessagePublication,
+	obsvReqC chan *gossipv1.ObservationRequest,
+	readiness readiness.Component,
+	chainID vaa.ChainID) *Watcher {
+
+	// // CosmWasm 1.0.0
+	// contractAddressFilterKey := "execute._contract_address"
+	// contractAddressLogKey := "_contract_address"
+
+	return &Watcher{
+		// urlWS:     urlWS,
+		urlLCD:    urlLCD,
+		contract:  contract,
+		msgChan:   lockEvents,
+		obsvReqC:  obsvReqC,
+		readiness: readiness,
+		chainID:   chainID,
+		// contractAddressFilterKey: contractAddressFilterKey,
+		// contractAddressLogKey:    contractAddressLogKey,
+		// latestBlockURL: latestBlockURL,
+	}
+}
+
+const TOPIC_LOG_MSG = "0x6eb224fb001ed210e379b335e35efe88672a8ce935d981a6896b27ffdf52a3b2"
+
+func (watcher *Watcher) Run(ctx context.Context) error {
+	networkName := vaa.ChainID(watcher.chainID).String()
+
+	p2p.DefaultRegistry.SetNetworkStats(watcher.chainID, &gossipv1.Heartbeat_Network{
+		ContractAddress: watcher.contract,
+	})
+
+	errC := make(chan error)
+	logger := supervisor.Logger(ctx)
+
+	// logger.Info("connecting to websocket", zap.String("network", networkName), zap.String("url", e.urlWS))
+
+	// c, _, err := websocket.DefaultDialer.DialContext(ctx, e.urlWS, nil)
+	// if err != nil {
+	// 	p2p.DefaultRegistry.AddErrorCount(e.chainID, 1)
+	// 	// connectionErrors.WithLabelValues(networkName, "websocket_dial_error").Inc()
+	// 	return fmt.Errorf("websocket dial failed: %w", err)
+	// }
+	// defer c.Close()
+
+	logger.Info("subscribed to new transaction events", zap.String("network", networkName))
+
+	readiness.SetReady(watcher.readiness)
+
+	// This function/thread queries and reports the current block height every interval
```

Comment:
I thought Hedera didn't have block numbers so we had to use timestamps? Or does the block query just return the timestamp as an int block number?

Reply:
Hedera does have block numbers